### ISSUE TITLE
Match Vision Highlight border to text unit resolution when reading text under the mouse

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#13136]
### Is your feature request related to a problem? Please describe.

Feature requested by a trainer (Dick via email) wanting more granular visual controls for their clients.

When a mouse user has the text unit resolution (NVDA mouse settings) set to character, word or line, moving the mouse over a block of text currently still results in the entire paragraph being highlighted.

### Describe the solution you'd like

Only the text NVDA will read should be read (the current, character, word, line or paragraph).

### Additional context
Tested current behaviour with NVDA 2021.3 rc1 in:

Firefox version: 94.0.2 
Chrome Version: 96.0.4664.45

**Comment by CyrilleB79** [comment]
More detailed steps to reproduce would be nice to have to understand better the problem.

It seems to me that the user has enabled "Follow mouse cursor" in "Review cursor" category (which is not the default). To me, it seems the only way to have paragraphs highlight moved when moving a mouse around a webpage.

Today:
* what is highlighted has nothing to do with text unit granularity: it is just the navigator object which is updated (along with the review cursor update) when the mouse is moved.
* the review cursor does not belong to the available highlighters

I guess that to satisfy this request, we should:
* add the review cursor as a new (fourth) possible highlighter
* take into account the granularity when moving the review cursor:
  * match text unit when moving the mouse
  * match text unit (char, word, line) when moving the review cursor with review cursor commands (numpad1 to 9).

A further step could be to match text unit when the browse mode cursor is displayed, according to the command being used: e.g. highlight the full word when control+left/rightArrow is used, etc.

**Comment by Qchristensen** [comment]
Good question, and evidently I didn't make detailed notes here, but just testing now, you are right, in order to have the paragraph highlighted (in red) you need to have "Follow mouse cursor" checked in NVDA's review cursor settings.  Without that checked, then often the entire window is highlighted - so having the paragraph highlighted is definitely closer to what the user requested here..  That is similar to what is highlighted when navigating the same text with the keyboard.  Potentially a solution to this issue would also resolve what is highlighted when you navigate text with say down arrow vs control+down arrow, vs right arrow, vs control+right arrow.  Currently when navigating by keyboard, a solid yellow border highlights the caret in Browse mode in the web browser (where there is no actual caret) - perhaps rather than introducing a new colour, this could be updated to highlight the current / most recently spoken text from that page - that may well be beyond the scope of this although when designing a solution, it would probably be worth considering both (let me know if a separate issue is needed for text navigation here).

**Comment by CyrilleB79** [comment]
Isn't there already a similar issue for keyboard navigation in text?

**Comment by Qchristensen** [comment]
> Isn't there already a similar issue for keyboard navigation in text?

Do you mean issue or feature?  I don't see another issue, but I have been known to miss them before - if anyone finds one, please do link to it here.

For feature (and what made me think twice about what you meant) - in browse mode, as you navigate through text, NVDA WILL highlight where the caret would be with a yellow outline, only around the current letter.

With the proposal in this issue, such a feature would ideally work with either keyboard or mouse navigation, and highlight as per the current resolution (so for mouse, if it is set to word, highlight the current word, but if moving by keyboard, if you press down arrow, highlight the whole line).

**Comment by Qchristensen** [comment]
Request for this from a blindness agency on behalf of clients.

**Comment by Adriani90** [comment]
cc: @hwf1324, @seanbudd this is still an issue in NVDA 64 bit last alpha right?

**Comment by hwf1324** [comment]
> cc: @hwf1324, @seanbudd this is still an issue in NVDA 64 bit last alpha right?

Yes, we have not made any changes to how we draw object positions at this time.

**Comment by hwf1324** [comment]
After a brief trial, this seems to be feasible. However, many edge cases need to be considered. This may require a lot of work.

If rectangles are to be displayed across multiple lines, we need to enable the highlighter to handle a list of rectangles for a single context, or calculate the union of a list of rectangles.
